rfcs/0048-BYOC-Marvell-ML-accelerator-integration.md

```diff
@@ -0,0 +1,547 @@
+  bypassing the tvm/Jenkinsfile's tests/scripts/task_rust.sh step. We will need help to re-enable the step.
+
+* We like to duplicate the Jenkins environment in order to run tvm/Jenkinsfile as is, but, we ran into many issues.
+  Currently, we have a tvm-like Jenksinsfile environment to only run a subset of test suites using a modified
+  Jenkinsfile.
+
+* We have identified a need to allow a call-back function to be registered when generating Mrvl-BYOC-specific
+  Nodes-JSON file. We are trying to follow TVM Python/CPP-CB style as much as possible. But, since our callback
+  function tvm/src/relay/backend/contrib/mrvl/graph_executor_codegen_mrvl.cc::GetExternalJSON() function is using
+  non-simple argument types, we need help from TVM community to provide suggestions/guidelines in order to make
+  new CB code better to meet TVM community requirements here.
+
+* For one Mrvl-BYOC relay transformation pass, we have identified a need to inject a (global) expr node ID for the
+  RelayExprNode class and its derived classes: Tuple and CallNode, so that during the transformation pass, we can
+  uniquely identify each Tuple or CallNode object. Again, we need help from TVM community to provide
+  suggestions/guidelines here in order to know whether this is one of the best ways to achieve the Mrvl-BYOC need.
+
+* We also identified a need to maintain linkages between (operator-)information described in the original, given
+  pre-trained network model and the code-gen JSON files so that the compiler backend will be able to report user-level
+  (e.g., meaningful-to-user) messages regarding the given pre-trained network. For instance, in the
+  tvm/python/tvm/relay/frontend/onnx.py and common.py files, we can see user-level information being captured using
+  “tvm_custom” related code as in original onnx.py file for the given pre-trained network; but, in common.py, the code
+  later drops the linkage, via attrs.pop(“tvm_custom”), and does not pass the linkage onto the initial relay IR graph.
+  We have a draft solution to maintain linkages between the given pre-trained network model and its relay IR graph
+  (using expr node ID and tvm custom ID, plus, a few utility functions), but would like to know whether the TVM
+  community has any better or work-in-progress resolution.
+
+* When using TVM RPC code to exercise and run inference on a remote-hosted Mrvl ML/AI HW accelerator for the Mrvl
+  subgraph, we ran into one minor issue and have made local TVM RPC enhancement so that, when a TVM RPC client sends
+  a file to the remote server, the TVM RPC client can know where the remote server saves the file on the remote machine.
+  Since this is not directly related to this Mrvl-BYOC PR, we will find time to contribute this enhance back in another
+  TVM PR soon.
+
+* In order for us to generate the constants-JSON file, we must “NOT” remove external params, which were stored in
```

Comment:
why is this? params passed in MetadataModule are meant for consumption only by the `runtime.Module` which defines them. it seems like perhaps you need to consume them at the executor level. could you explain that?

Reply:
We are using relay to generate JSON meta files representing the given network model in a way our (compiler-) backend code can process directly at compile time (e.g., only the marvell-part sub-graph(s). If we have include 100% of our backend code in the TVM codebase, then, we do not need to dump constants in JSON mega file; but, due to our backend code is built outside the typical TVM flow and can do other compile-time optimizations including manipulating constants, we need constants JSON.

Reply:
We will review RFC-#10 to find out. Thanks.